### DISCUSSION TITLE
Automate text insertion

### DISCUSSION THREAD

**Discussion by limaceous-bushwhacker** [gdh1995/vimium-c#1063] [Q&A] [answered]
How can I automate the following sequence?
1. Switch to tab 4 (`4g0`) ✅
2. Press Ctrl+I ❌
3. Insert input taken from Vomibar inside the text area opened by pressing <c-i> ❌
4. Press enter to search ❌

**Comment by gdh1995** [comment, marked as answer]
Well, it's quite complex, though indeed possible in Vimium C.

1. Firstly, you need a custom search engine from a keyword to `vimium://run/<runKey-commands-with-$s>`
2. Then the problem becomes to "how to write the `<runKey-commands-with-$s>`"
3. Its 1st command should be `4g0`
4. To do "press Ctrl+I", its 2nd command should be `dispatchEvent#ctrlKey&key=i`
- 1. `dispatchEvent` will returns a "failure" if there's no currently focused element, so `(` and `)` is necessary to ignore its result
5. To ensure the page has enough time to show a text area, you may need a 3rd command of `300wait` (to wait for an extra 300ms)
- 1. just a tip: a next command will run in 50ms by default, so `a+300wait+b` means to run `b` in `50+300+50=400ms`
6. To insert your query, a 4th command may be `editText##run=replace,$s`
- 1. the `replace` is a special action of the `editText` command, and designed to insert complex URL-encoded text
- 2. if you're sure that your query doesn't include space or comma, you may use `editText##run=exec,insertText,$S`
- 3. `##` means all following characters are a single parameter, so there's no place for the 5th command.

And, the final result is like:
```
g: vimium://run/4g0+(dispatchEvent#ctrlKey&key=i)+300wait+editText##run=replace,$s
```

**Reply by limaceous-bushwhacker** [reply]
This works for me. How can I execute (AKA press/send enter) the query?

EDIT: Nvm, `dispatchEvent#key=Enter` works :)

**Comment by gdh1995** [comment]
um, do you mean to simulate pressing enter after ctrl-i and simulate inputing?

```
dispatchEvent#key=Enter&keyCode=13
```

may help
